x/campaign/keeper/invariants.go

```diff
@@ -88,3 +91,48 @@ func VestingAccountWithoutCampaignInvariant(k Keeper) sdk.Invariant {
 		return "", false
 	}
 }
+
+// CampaignSharesInvariant invariant that checks if
+// the `MainnetVestingAccount` and `MainnetAccount` shares
+// sum is equal to existing campaign shares.
+func CampaignSharesInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		accounts := k.GetAllMainnetAccount(ctx)
+		vestingAccounts := k.GetAllMainnetVestingAccount(ctx)
+		shares := make(map[uint64]types.Shares)
+		for _, acc := range accounts {
+			if _, ok := shares[acc.CampaignID]; !ok {
+				shares[acc.CampaignID] = types.EmptyShares()
+			}
+			shares[acc.CampaignID] = types.IncreaseShares(
+				shares[acc.CampaignID],
+				acc.Shares,
+			)
+		}
+		for _, acc := range vestingAccounts {
+			if _, ok := shares[acc.CampaignID]; !ok {
+				shares[acc.CampaignID] = types.EmptyShares()
+			}
+			shares[acc.CampaignID] = types.IncreaseShares(
+				shares[acc.CampaignID],
+				acc.Shares,
+			)
+		}
+		for campaignID, share := range shares {
+			campaign, found := k.GetCampaign(ctx, campaignID)
+			if !found {
+				return sdk.FormatInvariant(
+					types.ModuleName, campaignShares,
+					fmt.Sprintf("%s: %d", types.ErrCampaignNotFound, campaignID),
+				), true
+			}
+			if !types.IsEqualShares(share, campaign.AllocatedShares) {
```

Comment:
We must also take into account the circulating vouchers.

We normally can query emitted tokens in `bank` from a denom with the `SupplyOf` query https://github.com/cosmos/cosmos-sdk/blob/1c468de77216e1da4c79c2a87a8e0ad307060185/x/bank/keeper/grpc_query.go#L90